File: 0893-all-nodes-distance-k-in-binary-tree/0893-all-nodes-distance-k-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    map<int, vector<int>> mp;
    vector<int> ans;
    vector<int> distanceK(TreeNode* root, TreeNode* target, int k) {
        dfs(root);
        bfs(target->val, k);
        return ans;
    }
    void dfs(TreeNode* root) {
        if(!root) return;
        if(root->left != NULL) {
            mp[root->val].push_back(root->left->val);
            mp[root->left->val].push_back(root->val);
        }
        if(root->right != NULL) {
            mp[root->val].push_back(root->right->val);
            mp[root->right->val].push_back(root->val);
        }
        dfs(root->left);
        dfs(root->right);
    }
    void bfs(int target, int k) {
        queue<int> q;
        q.push(target);
        int cnt = 0;
        unordered_set<int> vis;
        vis.insert(target);
        while(!q.empty()) {
            int n = q.size();
            if(cnt > k) break;
            for(int i = 0; i<n; i++) {
                int node = q.front();
                q.pop();
                if(cnt == k) ans.push_back(node);
                for(int i : mp[node]) {
                    if(!vis.count(i)) {
                        q.push(i);
                        vis.insert(i);
                    }
                }
            }
            cnt++;
        }
    }
};

File: 0893-all-nodes-distance-k-in-binary-tree/0893-all-nodes-distance-k-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> distanceK(TreeNode root, TreeNode target, int k) {
        Map<Integer, List<Integer>> map = new HashMap<>();
        dfs(root, map);
        List<Integer> ans = new ArrayList<>();
        Queue<Pair<Integer, Integer>> q = new LinkedList<>();
        q.add(new Pair(target.val, 0));
        Set<Integer> set = new HashSet<>();
        set.add(target.val);
        while(!q.isEmpty()) {
            var v = q.poll();
            if(v.getValue() == k) ans.add(v.getKey());
            if(v.getValue() < k) {
                for(int i : map.get(v.getKey())) {
                    if(!set.contains(i)) {  
                        q.add(new Pair(i, v.getValue() + 1));
                        set.add(i);
                    }
                }
            }
        }
        return ans;
    }
    public void dfs(TreeNode root, Map<Integer, List<Integer>> map) {
        if(root == null) return;
        map.putIfAbsent(root.val, new ArrayList<>());
        if(root.left != null) {
            map.putIfAbsent(root.left.val, new ArrayList<>());
            map.get(root.val).add(root.left.val);
            map.get(root.left.val).add(root.val);
        }
        if(root.right != null) {
            map.putIfAbsent(root.right.val, new ArrayList<>());
            map.get(root.val).add(root.right.val);
            map.get(root.right.val).add(root.val);
        }
        dfs(root.left, map);
        dfs(root.right, map);
    }
}